src/shared_utilities/checkpoints.jl

```diff
@@ -0,0 +1,128 @@
+import ClimaCore: InputOutput
+import ClimaUtilities
+
+"""
+    ClimaLand.find_restart(output_dir)
+
+Find the most recent restart file in the specified output directory.
+
+This function utilizes `ClimaUtilities.OutputPathGenerator.detect_restart_file`
+to locate the latest restart file within the output directory structure,
+assuming the `ActiveLinkStyle` is used for managing output folders.
+
+# Arguments
+- `output_dir`: The base output directory where the simulation results are stored.
+
+# Returns
+- The path to the most recent restart file found, or `nothing` if no restart
+  file is found.
+"""
+function find_restart(output_dir)
+    return ClimaUtilities.OutputPathGenerator.detect_restart_file(
+        ClimaUtilities.OutputPathGenerator.ActiveLinkStyle(),
+        output_dir,
+    )
+end
+
+"""
+    _context_from_Y(Y)
+
+Try extracting the context from the FieldVector Y.
+
+Typically Y has a structure like:
+```
+Y
+ .bucket
+        .T
+        .W
+        .Ws
+```
+`_context_from_Y` tries obtaining the context from the first Field in the
+hierarchy.
+"""
+function _context_from_Y(Y)
+    a_model_type = getproperty(Y, first(propertynames(Y)))
```

Comment:
ok! This works because we always do "Y.model.var"

Reply:
There's no guarantee that `Y.component.var` is a `Field`. One could have a scalar or a matrix there (we never do so, but also never forbid this)

Reply:
That's true. We make the state using coordinates, but the coordinates just need to be something we can iterate over. 

In that case, what would ClimaComms.context return? probably an error. Do we need a warning then?

Reply:
I was referring more to the possibility of us doing something like Y.canopy.energy.T (which actually, we do), in which case, this function would not work, right? Or maybe Y.canopy.energy is a field vector, and then it might still work

Reply:
Yes, I think `Y.canopy.energy` would not work if `energy` is a FieldVector.

Reply:
I updated this to be recursive so that it should find a context if there's any Field in the hierarchy.

Reply:
nice!!